lib/collection/src/shards/transfer/driver.rs

```diff
@@ -68,6 +75,40 @@ pub async fn transfer_shard(
             )
             .await?;
         }
+
+        // Attempt to transfer WAL delta
+        ShardTransferMethod::WalDelta => {
+            let result = transfer_wal_delta(
+                transfer_config,
+                shard_holder.clone(),
+                shard_id,
+                remote_shard.clone(),
+                channel_service,
+                consensus,
+                collection_name,
+            )
+            .await;
+
+            // If WAL delta transfer failed, fall back to stream records
+            // TODO: fall back to preferred transfer method in config, snapshot transfer may be
+            // TODO: problematic if it expects a different shard replica set state
+            // TODO: let method = self.shared_storage_config
+            // TODO:     .default_shard_transfer_method
+            // TODO:     .unwrap_or_default();
+
+            if let Err(err) = result {
+                log::warn!(
+                    "Failed to do shard diff transfer, falling back to stream records: {err}"
+                );
+                transfer_stream_records(
+                    shard_holder.clone(),
+                    shard_id,
+                    remote_shard,
+                    collection_name,
+                )
+                .await?;
+            }
```

Comment:
This is a naive fallback approach. I'll improve this in a follow up PR, and it is noted in our tracking issue.

Reply:
This is currently broken because it leaves the recovering node in unexpected state. Our stream records transfer logic cannot continue just on that.

Reply:
I temporarily fixed this by switching the shard state from `PartialSnapshot` back to `Partial`, required for stream records transfer. But it's more a hack than a right way of doing things.